drivers/leds/ws2812.c

```diff
@@ -55,15 +55,7 @@
  * Pre-processor Definitions
  ****************************************************************************/
 
-#ifdef WS2812_HAS_WHITE
-#  define WS2812_RW_PIXEL_SIZE  4
-#else 
-#  define WS2812_RW_PIXEL_SIZE  3
-#endif
-
-#ifdef CONFIG_WS2812_NON_SPI_DRIVER
-
-#else /* CONFIG_WS2812_NON_SPI_DRIVER */
```

Comment:
@tmedicci I think there are some models of WS2812 that doesn't have WHITE LED, so instead of using 32-bit it uses 24-bit.

@vbenso could you please confirm/verify?

Reply:
Hi @acassis !

Yes, I know (in fact, the "real" ws2812 doesn't contain 4 LEDs at all, only variants that use the same protocol and sometimes are sold as ws2812 RGBW), but the variable that the LED driver uses is 4-byte long (independent of the LED). The lower-half driver should handle that. I explained in the commit message:

> Although the LED might be RGB-only, the LED data is packed in a 32-bit long variable and, then, this is the default size of a LED pixel to define the 'WS2812_RW_PIXEL_SIZE' macro. Please note that the lower-half driver will deal with the case of the addressable LED being 3 or 4-pixel sized.

Reply:
@acassis , just to make it clearer: the model is 24-bit, but this is handled by the lower-half driver. The way the upper-half driver works on NuttX, it uses 32-bit (a word) for both RGB and RGBW models.

Reply:
Hi guys, merry Christmas.
I didn't want to allocate 32 bits unless it was strictly required. That being said, I understand that using packed structures could degrade the performance. One could say that there is plenty of memory and people shouldn't care too much about it. In my case, I was using 700+ leds, so there were some memory concerns.

Reply:
> Hi guys, merry Christmas. I didn't want to allocate 32 bits unless it was strictly required. That being said, I understand that using packed structures could degrade the performance. One could say that there is plenty of memory and people shouldn't care too much about it. In my case, I was using 700+ leds, so there were some memory concerns.

Hi, Merry Christmas!

Yes, I understand your concerns about using a 32-bit when only 24-bit is necessary. However, no application that used the WS2812 encodes the LED pixel data in 24-bit space (this would depend on the application itself). 
Even considering that the application could encode LED data in 24-bit/pixel, we still had:

```
#ifdef WS2812_HAS_WHITE
#  define WS2812_RW_PIXEL_SIZE  4
#else 
#  define WS2812_RW_PIXEL_SIZE  3
#endif
```
 with `WS2812_HAS_WHITE` not being defined by Kconfig (it should be 'CONFIG_WS2812_HAS_WHITE`). Also, the [`ws2812_write`](https://github.com/apache/nuttx/blob/d1326e81bca47617aa253705ca778c7d51f82038/drivers/leds/ws2812.c#L316) (for `CONFIG_WS2812_NON_SPI_DRIVER`) is not compliant with raw 24-bit data length (LED pixel data should be encoded in 32-bit space).